### DISCUSSION TITLE
VBx OERN usage pattern fixes

### DISCUSSION THREAD

**Discussion by mburns08109** [twinbasic/twinbasic#1246] [General]
### ON ERROR RESUME NEXT  <-- A NECESSARY evil - or never necessary?

Elsewhere here, dmrvb proffered this as a "legit" use case for an OERN:
```
On Error Resume Next
For Each F in Forms
call "method that only some forms have"
Next
```

I join critics in saying it isn't, except that is it due to some missing VBx language capabilities that really should be there in the language - and not only available via arcane API calls to get the answer right.

We need a DIRECT language construct to cleanly support a queryinterface function to TEST whether a given class instance variable supports a named method - or not - and if so, what interface NAME/Type that method is supported on.

Something like a built-in:
```
public function SupportsMethod(byref o as object, byval MethodName as string, byref TypeName as string) as boolean
...
end function
```

So in the above snippet that we could directly TEST each F to see if it supports "method that only some forms have" - and react accordingly with some code pattern like:

```
if SupportsMethod(F, "method that only some forms have", strNewType as string) then
  if len(strNewType) > 0 then set f = cTypeName(F, strNewType)   ' changes/casts F to type indicated by string name (always a support interface since SupportsMethod returns it based on what F actually implements)
  Call F.["method that only some forms have"]
'else    'not needed if we're just skipping this case
endif
```

Naturally, we'd need to deal with differing method signatures and return type issues with some appropriate semantics as well here, but without something like this being natively available as a language construct, I can't see a way to totally eliminate the need for OERN-type constructs like DMRVB illustrated.

**Comment by FullValueRider** [comment]
The provided code is better written as as try function.

https://rubberduckvba.wordpress.com/2019/05/

**Reply by mburns08109** [reply]
Aside from the entirely relevant point that not everyone has or uses (or _can_ use due to licensing/restrictions/etc.) rubberduck, there is still the need for a QueryInterface functional construct in tB _without the need to resort to API calls_. This is more than a syntactic sugar thing - it is an important part of the COM infrastructure that should have been exposed in VBx loooong ago, but wasn't.

**Reply by wqweto** [reply]
I think that `TypeOf ... Is` operator is a fairly complete reimplementation (in BASIC-y way) of COM's QueryInterface functional construct.

**Reply by mburns08109** [reply]
Be that as it may, we need a stronger version so we can check on method names on the interface(s).

**Reply by FullValueRider** [reply]
@mburns08109 The Rubberduck blog reference refers to a coding pattern that eliminates the issues with OERN.  I agree, it would be nice to be able to query an object for a specific method.

**Comment by mburns08109** [comment]
Perhaps what is needed is a series of (intrinsic, COM) functions:
SupportsMethod() as boolean
SupportsInterface() as boolean
SupportsSignature() as boolean
...along these lines - the function names being obvious hints as to their purposes.
- inquire as to supported method name
- inquire as to supported interface(/type)
- inqure as to supported method signature(s?)

My question from this last thought might be: would that last one not require some new meta-type datatype (to describe method signature metadata)?

This stuff is what the VBx language has been missing.

I should mention: the PURPOSE of these functional language extensions is to rid us of the OERN coding pattern requirement.
That pattern exists largely to fill the gaps present by our inability - as a defined function of the VBx language (as opposed to via Win32 API calls) - to query OBJECT type variable meta-data on interfaces(/Types) and their methods, and method signatures.

**Reply by mburns08109** [reply]
I had an insight earlier on a new version of the COM specification, too:

Change this one basic COM interface definition semantic from:

All method signatures must be static on an interface

to

All method signatures are static but are considered to be expandable (i.e. additional new parameters MAY be added in the future, but newly defined, added parameters for a version of a component are then considered static requirements for all new future versions going forward).

Call it the COMOpen spec?
